src/progress.rs

```diff
@@ -50,6 +53,92 @@ impl<'a> Progress<'a> {
         )
         .await
     }
+
+    /// Wait until MeiliSearch processes an update, and get its status.
+    /// 
+    /// interval_ms = The frequency at which the server should be polled in milliseconds. Default = 50ms
+    /// timeout_ms = The maximum time to wait for processing to complete in milliseconds. Default = 5000ms
+    /// 
+    /// # Example
+    ///
+    /// ```
+    /// # use meilisearch_sdk::{client::*, document, indexes::*, progress::*};
+    /// # use serde::{Serialize, Deserialize};
+    /// 
+    /// # #[derive(Debug, Serialize, Deserialize, PartialEq)]
+    /// # struct Document {
+    /// #    id: usize,
+    /// #    value: String,
+    /// #    kind: String,
+    /// # }
+    /// # 
+    /// # impl document::Document for Document {
+    /// #    type UIDType = usize;
+    /// #
+    /// #    fn get_uid(&self) -> &Self::UIDType {
+    /// #        &self.id
+    /// #    }
+    /// # }
+    /// 
+    /// # futures::executor::block_on(async move {
+    /// let client = Client::new("http://localhost:7700", "masterKey");
+    /// let movies = client.create_index("movies_wait_for_pending", None).await.unwrap();
+    /// 
+    /// let progress = movies.add_documents(&[
+    ///     Document { id: 0, kind: "title".into(), value: "The Social Network".to_string() },
+    ///     Document { id: 1, kind: "title".into(), value: "Harry Potter and the Sorcerer's Stone".to_string() },
+    /// ], None).await.unwrap();
+    /// 
+    /// let status = progress.wait_for_pending_update(None, None).await.unwrap();
+    /// 
+    /// # client.delete_index("movies_wait_for_pending").await.unwrap();
+    /// assert!(matches!(status, UpdateStatus::Processed { .. }));
+    /// # });
+    /// ```
+    pub async fn wait_for_pending_update(
+        &self,
+        interval_ms: Option<u64>,
+        timeout_ms: Option<u64>,
+    ) -> Result<UpdateStatus, Error> {
+        let interval: u64;
+        let timeout: u64;
+
+        match interval_ms {
+            Some(v) => interval = v,
+            None => interval = 50,
+        }
+
+        match timeout_ms {
+            Some(v) => timeout = v,
+            None => timeout = 5000,
+        }
+
+        let mut elapsed_time = 0;
+        let mut status: UpdateStatus;
+
+        while timeout > elapsed_time {
+            status = self.get_status().await?;
+
+            match status {
+                UpdateStatus::Failed { .. } | UpdateStatus::Processed { .. } => {
+                    return self.get_status().await;
+                },
+                UpdateStatus::Enqueued { .. } => {
+                    elapsed_time += interval;
+                    sleep(Duration::from_millis(interval));
```

Comment:
This sleep function is blocking. It should be async since this function is async. It will be hard without adding a new dependency... On wasm, we can use `wasm-bindgen-futures`. Do you have any idea for native?

Reply:
Good point. I’m not sure of a way to do this with the standard library either.

Reply:
I've done a bit more research on this and the only solutions I am finding are to use the sleep from either the async_std crate or tokio crate.

Reply:
You are right. I think it would be extremely difficult without dependency. `futures` is smaller than `async_std` and `tokio` so let's use `futures`. The dependency should be added in the `[target.'cfg(not(target_arch = "wasm32"))'.dependencies]` section of `Cargo.toml`, near `isahc`. Then you will have to use `cfg`s to create a custom async sleep function.

On native:
- Spawn a future
- Create a thread which owns an async oneshot channel sender
- Asynchronously wait for data in the receiver part of the channel, outside the thread

On wasm:
Copy and improve [this sleep function](https://github.com/Mubelotix/wasm-extension-template/blob/master/src/util.rs). `js-sys` dependency will need to be added but it's already used by web-sys so the executable size should not be increased.

Reply:
This is what I have so far for the native. It is passing the tests, but this is also my first time using the Futures crate so I have a few questions.

```rust
use std::thread::sleep;
use std::time::Duration;
use futures::channel::oneshot::{channel, Sender};
...

pub async fn wait_for_pending_update(
        &self,
        interval_ms: Option<Duration>,
        timeout_ms: Option<Duration>,
    ) -> Result<UpdateStatus, Error> {
    ...
                UpdateStatus::Enqueued { .. } => {
                    elapsed_time += interval;
                    let (sender, receiver) = channel::<()>();
                    let _thread = std::thread::spawn(move || async_sleep(sender, interval));
                    receiver.await.unwrap();
                },
    ...
}

#[cfg(not(target_arch = "wasm32"))]
pub(crate) fn async_sleep(sender: Sender<()>, interval: Duration) {
    sleep(interval);
    sender.send(()).unwrap();
}
```

1. Does this look correct, or would you expect something different?
2. Is it enough to send the sleep into the channel, or do I need to send the whole loop?
2. Do you know of a way I can confirm that this isn't blocking?

Reply:
Great ! It looks correct and that's what I was expecting. It suggest you move the whole workaround in the async_sleep function, including the thread so that you get a nicer interface usable like this :
```rust
UpdateStatus::Enqueued { .. } => {
    elapsed_time += interval;
    async_sleep(sender, interval).await;
},
```
You will also need to move the specific `use` keywords in the function since wasm will have no access to the `future` dependency.

You could do some logging in multiple futures `join!`ed together to confirm that it is not blocking but it should be visible in the code. The only blocking function called is wrapped in the new thread so the main thread cannot be blocked in any way.

Reply:
After making this update so it can be called with `async_sleep(interval).await;` it is now:

```rust
#[cfg(not(target_arch = "wasm32"))]
pub(crate) async fn async_sleep(interval: Duration) {
    let (sender, receiver) = futures::channel::oneshot::channel::<()>();
    let _thread = std::thread::spawn(move || std::thread::sleep(interval));
    sender.send(()).unwrap();
    let _ = async { receiver.await };
}
```

This function and the wasm one are currently in the progress.rs file, but even though they are currently only used here it seems possible they could be used in other places in the future. With that should they be moved to a different location? Maybe a sleep.rs file, utils.rs file, or something like this?

Reply:
That's a lot cleaner thanks! You are right about moving this function; that could be useful in the future. However, I doubt there will be other usages in the crate. Maybe you could even bury the function inside `wait_for_pending_update`. Anyway, we will remember the existence of this code and will use it if needed.
Feel free to ask for help while building the Wasm version.

There is a little problem with this code. We have to send data through the channel **from** the new thread, after the sleep.

Reply:
For the wasm version I currently have:

```rust
#[cfg(target_arch = "wasm32")]
pub(crate) async fn async_sleep(interval: Duration) {
    use wasm_bindgen_futures::JsFuture;
    
    JsFuture::from(js_sys::Promise::new(&mut |yes, _| {
        web_sys::window()
            .unwrap()
            .set_timeout_with_callback_and_timeout_and_arguments_0(
                &yes,
                interval.as_millis() as i32,
            )
            .unwrap();
    })).await.unwrap();
}
```

When I try to run the tests to see what needs updating with `cargo test --target=wasm32-unknown-unknown -- --test-threads=1` I get an error `cannot execute binary file` (this error happens even without the new function present). I have found other people getting this error, but no one with a solution. Any ideas?

Reply:
There is no way to test the wasm executable out of the box. You have to call the code from a complete wasm project to test things. I have made [an example](https://github.com/meilisearch/meilisearch-rust/tree/main/examples/web_app) of using MeiliSearch in wasm. It should be a great resource to start since everything is ready to be compiled and served.